### ISSUE TITLE
Reuse test prefixes instead of rebuilding them for each test

### ISSUE THREAD

**Issue by hosef** [backdrop/backdrop-issues#4353]
**Description of the need**

Currently, when a test is done it deletes all the tables and files for it's prefix. During most tests, only a few tables are modified, which means that most of the table dropping and creating is unnecessary.
**Proposed solution**

Change the setUp and tearDown functions from the test class to not delete the class prefix, but instead replace the tables that were changed.

**Comment by docwilmot** [comment]
Whats the expected advantage? Speed? I tested it and it was not faster, and seemed to be almost certainly slower on the one test I tried (Registration Module Test, from the newly ported Registration module. 11m 40s with PR but 8m 48s with default 1.x.

**Comment by hosef** [comment]
I was trying to make it faster. I would like to know how your environment is setup and how you are running your tests, because that is the opposite of what I experienced when I was running tests through the testing UI on my local environment.

**Comment by docwilmot** [comment]
Wampserver (WAMP) on Windows 10. PHP 5.6. Did you try with that particular module test?

**Comment by indigoxela** [comment]
Wow, I totally missed this issue. Deleting only the changed tables could be a significant speed improvement, especially on newer environments, where the conversion to MyISAM fails because of the index key length problem. @hosef how can I help?

**Comment by hosef** [comment]
@indigoxela I think I have a working version of the PR finally. It would be good to see how the performance of my PR compares with stock Backdrop on various environments.

**Comment by hosef** [comment]
Because the test boxes are using MyISAM currently, there is very low overhead from copying tables, so there will be almost no effect there.

**Comment by hosef** [comment]
Hmm, seems there is still a problem with locking. :|

**Comment by indigoxela** [comment]
There are a lot of failing tests.

@hosef mind to add a `Fixes [link to this issue]` to your PR main comment? It's easier to find here, if it's linked in the issue header.

**Comment by hosef** [comment]
Yeah, it seems that multiple test runners are still selecting the same prefix. I added a lock_wait in there, but they appear to be ignoring it, so I am not sure what to do about that.

**Comment by indigoxela** [comment]
To be honest, I don't fully understand your approach, so please forgive if I write something silly... :grin: 

You add two new columns to table "simpletest_prefix" - "profile" and "in_use", and based on these you try to pick a copy of the requested profile that's currently *not* in use by a test runner - is that correct so far?

But they seem to clash? Based on the exceptions I see in the failed test runs, it could also be that they pick the wrong profile (tests don't find expected tables or the table already exists when they try to create). Wild guessing without deeper understanding...

Currently "only" 22 (28) tests are failing. If it were a base concept problem, I'd actually expect more failures.

**Comment by hosef** [comment]
> You add two new columns to table "simpletest_prefix" - "profile" and "in_use", and based on these you try to pick a copy of the requested profile that's currently _not_ in use by a test runner - is that correct so far?
Mostly yes. I also check that the test_id is set to 0.


> But they seem to clash? Based on the exceptions I see in the failed test runs, it could also be that they pick the wrong profile (tests don't find expected tables or the table already exists when they try to create). Wild guessing without deeper understanding...
As far as I can tell, those errors are coming from BackdropWebTestCase::useCache(). When 2 of the runners select the same test prefix, they will both try to delete and recreate the same tables. One of them will succeed with the table deletion and the other will fail and stop processing.

> Currently "only" 22 (28) tests are failing. If it were a base concept problem, I'd actually expect more failures.
The number of failures changes each time that the tests are run. It is essentially a race between each runner as to which gets to the setup step first.
I added a few things to try and keep them spaced out a little bit, such as the lock_wait() in prepareDatabasePrefix() and the usleep() in run-tests.sh, but they don't appear to be working. I'm at bit of a loss as to what it might take to fix the race conditions at this point.
